In [1]:
import pandas as pd
import numpy as np

import matplotlib.pylab as plt
import seaborn as sns

import env

## Attendance Data

### Read the data from the attendance table and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [2]:
url = env.get_db_url('tidy_data')
query = 'SELECT * FROM attendance'

df = pd.read_sql(query, url)
df

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [3]:
df = df.rename(columns={'Unnamed: 0':'name'})
df

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [4]:
df = df.melt(id_vars='name', var_name='date', value_name='attend')
df

name        date attend
0   Sally  2018-01-01      P
1    Jane  2018-01-01      A
2   Billy  2018-01-01      A
3    John  2018-01-01      P
4   Sally  2018-01-02      T
5    Jane  2018-01-02      P
6   Billy  2018-01-02      T
7    John  2018-01-02      T
8   Sally  2018-01-03      T
9    Jane  2018-01-03      T
10  Billy  2018-01-03      A
11   John  2018-01-03      H
12  Sally  2018-01-04      H
13   Jane  2018-01-04      T
14  Billy  2018-01-04      A
15   John  2018-01-04      P
16  Sally  2018-01-05      P
17   Jane  2018-01-05      T
18  Billy  2018-01-05      H
19   John  2018-01-05      P
20  Sally  2018-01-06      A
21   Jane  2018-01-06      T
22  Billy  2018-01-06      T
23   John  2018-01-06      T
24  Sally  2018-01-07      T
25   Jane  2018-01-07      A
26  Billy  2018-01-07      P
27   John  2018-01-07      P
28  Sally  2018-01-08      T
29   Jane  2018-01-08      T
30  Billy  2018-01-08      T
31   John  2018-01-08      P

In [5]:
df.attend = df.attend.map({'P':1, 'H':.5, 'A':0, 'T':.9})
df

name        date  attend
0   Sally  2018-01-01     1.0
1    Jane  2018-01-01     0.0
2   Billy  2018-01-01     0.0
3    John  2018-01-01     1.0
4   Sally  2018-01-02     0.9
5    Jane  2018-01-02     1.0
6   Billy  2018-01-02     0.9
7    John  2018-01-02     0.9
8   Sally  2018-01-03     0.9
9    Jane  2018-01-03     0.9
10  Billy  2018-01-03     0.0
11   John  2018-01-03     0.5
12  Sally  2018-01-04     0.5
13   Jane  2018-01-04     0.9
14  Billy  2018-01-04     0.0
15   John  2018-01-04     1.0
16  Sally  2018-01-05     1.0
17   Jane  2018-01-05     0.9
18  Billy  2018-01-05     0.5
19   John  2018-01-05     1.0
20  Sally  2018-01-06     0.0
21   Jane  2018-01-06     0.9
22  Billy  2018-01-06     0.9
23   John  2018-01-06     0.9
24  Sally  2018-01-07     0.9
25   Jane  2018-01-07     0.0
26  Billy  2018-01-07     1.0
27   John  2018-01-07     1.0
28  Sally  2018-01-08     0.9
29   Jane  2018-01-08     0.9
30  Billy  2018-01-08     0.9
31   John  2018-01-08     1.0

In [6]:
df.groupby('name').mean()

attend
name         
Billy  0.5250
Jane   0.6875
John   0.9125
Sally  0.7625

### You should end up with something like this:

```python
name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64
```

## Coffee Levels

### Read the coffee_levels table.

In [7]:
query = 'SELECT * FROM coffee_levels'

df = pd.read_sql(query, url)
df.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

### Transform the data so that each carafe is in it's own column.

In [8]:
df.pivot_table(index='hour', columns='coffee_carafe')

coffee_amount                    
coffee_carafe             x         y         z
hour                                           
8                  0.816164  0.189297  0.999264
9                  0.451018  0.521502  0.915990
10                 0.843279  0.023163  0.144928
11                 0.335533  0.235529  0.311495
12                 0.898291  0.017009  0.771947
13                 0.310711  0.997464  0.398520
14                 0.507288  0.058361  0.864464
15                 0.215043  0.144644  0.436364
16                 0.183891  0.544676  0.280621
17                 0.391560  0.594126  0.436677

### Is this the best shape for the data?

#### No

In [9]:
df

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

## Cake Recipes

### Read the cake_recipes table. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [10]:
query = 'SELECT * FROM cake_recipes'

df = pd.read_sql(query, url)
df

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

### Tidy the data as necessary.

In [11]:
df = df.melt(id_vars='recipe:position', var_name='temp', value_name='score')
df

recipe:position temp      score
0         a:bottom  225  61.738655
1            a:top  225  51.709751
2         b:bottom  225  57.095320
3            b:top  225  82.455004
4         c:bottom  225  96.470207
5            c:top  225  71.306308
6         d:bottom  225  52.799753
7            d:top  225  96.873178
8         a:bottom  250  53.912627
9            a:top  250  52.009735
10        b:bottom  250  61.904369
11           b:top  250  95.224151
12        c:bottom  250  52.001358
13           c:top  250  82.795477
14        d:bottom  250  58.670419
15           d:top  250  76.101363
16        a:bottom  275  74.414730
17           a:top  275  68.576858
18        b:bottom  275  61.196980
19           b:top  275  98.594881
20        c:bottom  275  92.893227
21           c:top  275  92.098049
22        d:bottom  275  51.747686
23           d:top  275  59.571620
24        a:bottom  300  98.786784
25           a:top  300  50.228470
26        b:bottom  300  99.248541
27           b:top  300  58.169349
28        c:bottom  300  65.473084
29           c:top  300  53.960273
30        d:bottom  300  56.183110
31           d:top  300  50.971626

In [12]:
df[['recipe','position']] = df['recipe:position'].str.split(':', expand=True)
df

recipe:position temp      score recipe position
0         a:bottom  225  61.738655      a   bottom
1            a:top  225  51.709751      a      top
2         b:bottom  225  57.095320      b   bottom
3            b:top  225  82.455004      b      top
4         c:bottom  225  96.470207      c   bottom
5            c:top  225  71.306308      c      top
6         d:bottom  225  52.799753      d   bottom
7            d:top  225  96.873178      d      top
8         a:bottom  250  53.912627      a   bottom
9            a:top  250  52.009735      a      top
10        b:bottom  250  61.904369      b   bottom
11           b:top  250  95.224151      b      top
12        c:bottom  250  52.001358      c   bottom
13           c:top  250  82.795477      c      top
14        d:bottom  250  58.670419      d   bottom
15           d:top  250  76.101363      d      top
16        a:bottom  275  74.414730      a   bottom
17           a:top  275  68.576858      a      top
18        b:bottom  275  61.196980      b   bottom
19           b:top  275  98.594881      b      top
20        c:bottom  275  92.893227      c   bottom
21           c:top  275  92.098049      c      top
22        d:bottom  275  51.747686      d   bottom
23           d:top  275  59.571620      d      top
24        a:bottom  300  98.786784      a   bottom
25           a:top  300  50.228470      a      top
26        b:bottom  300  99.248541      b   bottom
27           b:top  300  58.169349      b      top
28        c:bottom  300  65.473084      c   bottom
29           c:top  300  53.960273      c      top
30        d:bottom  300  56.183110      d   bottom
31           d:top  300  50.971626      d      top

In [13]:
df = df.drop(columns='recipe:position')
df

temp      score recipe position
0   225  61.738655      a   bottom
1   225  51.709751      a      top
2   225  57.095320      b   bottom
3   225  82.455004      b      top
4   225  96.470207      c   bottom
5   225  71.306308      c      top
6   225  52.799753      d   bottom
7   225  96.873178      d      top
8   250  53.912627      a   bottom
9   250  52.009735      a      top
10  250  61.904369      b   bottom
11  250  95.224151      b      top
12  250  52.001358      c   bottom
13  250  82.795477      c      top
14  250  58.670419      d   bottom
15  250  76.101363      d      top
16  275  74.414730      a   bottom
17  275  68.576858      a      top
18  275  61.196980      b   bottom
19  275  98.594881      b      top
20  275  92.893227      c   bottom
21  275  92.098049      c      top
22  275  51.747686      d   bottom
23  275  59.571620      d      top
24  300  98.786784      a   bottom
25  300  50.228470      a      top
26  300  99.248541      b   bottom
27  300  58.169349      b      top
28  300  65.473084      c   bottom
29  300  53.960273      c      top
30  300  56.183110      d   bottom
31  300  50.971626      d      top

### Which recipe, on average, is the best? recipe b

In [14]:
df.groupby('recipe').mean().idxmax()

score    b
dtype: object

In [15]:
df.groupby('recipe').mean().max()

score    76.736074
dtype: float64

### Which oven temperature, on average, produces the best results? 275

In [16]:
df.groupby('temp').mean().sort_values('score').tail(1)

score
temp           
275   74.886754

In [17]:
df.groupby('temp').mean().max()

score    74.886754
dtype: float64

### Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [18]:
df.groupby(['recipe','position','temp']).mean().idxmax()

score    (b, bottom, 300)
dtype: object

In [19]:
df.groupby(['recipe','position','temp']).mean().max()

score    99.248541
dtype: float64

In [20]:
df.groupby(['recipe', 'position', 'temp']).mean().sort_values('score').tail(1)

score
recipe position temp           
b      bottom   300   99.248541